File: _CodeTopics/LeetCode/201-400/000203/tran/000203_impl.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {

        struct ListNode* dummyHead = new ListNode(0);
        dummyHead->next = head;
        struct ListNode* prev = dummyHead;

        while (prev->next) {
            if (prev->next->val != val) {
                prev = prev->next;
            } else {
                prev->next = prev->next->next;
            }
        }
        return dummyHead->next;
    }
};

/*
https://leetcode-cn.com/submissions/detail/184194128/

66 / 66 个通过测试用例
状态：通过
执行用时: 28 ms
内存消耗: 14.6 MB

执行用时：28 ms, 在所有 C++ 提交中击败了68.56%的用户
内存消耗：14.6 MB, 在所有 C++ 提交中击败了56.59%的用户
*/

File: _CodeTopics/LeetCode/201-400/000203/000203_impl.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """

        # 如果肯用个dummyHead（并且while循环里不是用curr判断，而是用prev.next来判断！），
        # 就不用单独考虑初始的head（以及之后的若干节点）不能用的问题了。

        dummyHead = ListNode(0)
        dummyHead.next = head
        prev = dummyHead
       
        while prev.next:
            if prev.next.val != val:
                prev = prev.next
            else:
                prev.next = prev.next.next
        return dummyHead.next
        
"""
https://leetcode-cn.com/submissions/detail/184181461/

66 / 66 个通过测试用例
状态：通过
执行用时: 68 ms
内存消耗: 19.8 MB

执行用时：68 ms, 在所有 Python 提交中击败了17.40%的用户
内存消耗：19.8 MB, 在所有 Python 提交中击败了28.89%的用户
"""
